### PR TITLE
Introduce TokenHelpers

### PR DESCRIPTION
**Pull request by Maxwe11**
- IsFirstInLine
- IsLastInLine
- IsPrecededBySpace
- IsFollowedBySpace

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TokenHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            SyntaxToken nextToken = token.GetNextToken();
+            return nextToken.IsKind(SyntaxKind.None) || token.GetLine() < nextToken.GetLine();
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the <paramref name="token"/> preceded by a space.
+        /// </summary>
+        /// <param name="token">The token to process.</param>
+        /// <returns>true if token is preceded by a space, otherwise false.</returns>
+        internal static bool IsPrecededBySpace(this SyntaxToken token)
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TokenHelpers.cs]: I would prefer to call the method IsPrecededByWhitespace, as that more accurately describes the functionality.

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TokenHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+            SyntaxTriviaList triviaList = token.GetPreviousToken().TrailingTrivia.AddRange(token.LeadingTrivia);
+            return triviaList.Count > 0 && triviaList.Last().IsKind(SyntaxKind.WhitespaceTrivia);
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the <paramref name="token"/> followed by a space.
+        /// </summary>
+        /// <param name="token">The token to process.</param>
+        /// <returns>true if token is followed by a space, otherwise false.</returns>
+        internal static bool IsFollowedBySpace(this SyntaxToken token)
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TokenHelpers.cs]: I would prefer `IsFollowedByWhitespace` as that more accurately describes the functionality.

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1009ClosingParenthesisMustBeSpacedCorrectly.cs [hunk 3]
```diff
@@ -78,26 +79,12 @@ private void HandleCloseParenToken(SyntaxTreeAnalysisContext context, SyntaxToke
                 return;
             }
 
-            bool precededBySpace;
-
-            bool followedBySpace;
-            bool lastInLine;
+            bool precededBySpace = token.IsFirstInLine() ? true : token.IsPrecededBySpace();
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1009ClosingParenthesisMustBeSpacedCorrectly.cs]: You could also write this as `precededBySpace = token.IsFirstInLine() || token.IsPrecededBySpace();`

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1011ClosingSquareBracketsMustBeSpacedCorrectly.cs [hunk 4]
```diff
@@ -83,26 +84,12 @@ private void HandleCloseBracketToken(SyntaxTreeAnalysisContext context, SyntaxTo
                 return;
             }
 
-            bool precededBySpace;
-            bool firstInLine;
-
-            bool followedBySpace;
-            bool lastInLine;
+            bool firstInLine = token.IsFirstInLine();
+            bool precededBySpace = firstInLine ? true : token.IsPrecededBySpace();
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1011ClosingSquareBracketsMustBeSpacedCorrectly.cs]: You could also write it like `precededBySpace = firstInLine || token.IsPrecededBySpace();`

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1009ClosingParenthesisMustBeSpacedCorrectly.cs (hunk 3 above)
Comment (Maxwe11) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1009ClosingParenthesisMustBeSpacedCorrectly.cs]: nice catch, will fix all occurrences

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1013ClosingCurlyBracketsMustBeSpacedCorrectly.cs [hunk 5]
```diff
@@ -77,26 +78,12 @@ private void HandleCloseBraceToken(SyntaxTreeAnalysisContext context, SyntaxToke
                 return;
             }
 
-            bool precededBySpace;
-            bool firstInLine;
-
-            bool followedBySpace;
-            bool lastInLine;
+            bool firstInLine = token.IsFirstInLine();
+            bool precededBySpace = firstInLine ? true : token.IsPrecededBySpace();
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1013ClosingCurlyBracketsMustBeSpacedCorrectly.cs]: You could also write this as `precededBySpace = firstInLine || token.IsPrecededBySpace();`

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1015ClosingGenericBracketsMustBeSpacedCorrectly.cs [hunk 6]
```diff
@@ -86,27 +87,13 @@ private void HandleGreaterThanToken(SyntaxTreeAnalysisContext context, SyntaxTok
...
             }
 
-            bool precededBySpace;
-            bool firstInLine;
-
-            bool followedBySpace;
-            bool lastInLine;
+            bool firstInLine = token.IsFirstInLine();
+            bool lastInLine = token.IsLastInLine();
+            bool precededBySpace = firstInLine ? true : token.IsPrecededBySpace();
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1015ClosingGenericBracketsMustBeSpacedCorrectly.cs]: You could also write this like `precededBySpace = firstInLine || token.IsPrecededBySpace();`

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1019MemberAccessSymbolsMustBeSpacedCorrectly.cs [hunk 7]
```diff
@@ -96,23 +97,9 @@ private void HandleQuestionToken(SyntaxTreeAnalysisContext context, SyntaxToken
...
-            }
-            else
-            {
-                SyntaxToken precedingToken = token.GetPreviousToken();
-                precededBySpace = precedingToken.HasTrailingTrivia;
-            }
-
-            followedBySpace = token.HasTrailingTrivia;
+            bool firstInLine = token.IsFirstInLine();
+            bool precededBySpace = firstInLine ? true : token.IsPrecededBySpace();
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1019MemberAccessSymbolsMustBeSpacedCorrectly.cs]: You could also write this like `precededBySpace = firstInLine || token.IsPrecededBySpace();`

File: StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1023DereferenceAndAccessOfSymbolsMustBeSpacedCorrectly.cs [hunk 8]
```diff
@@ -121,22 +122,10 @@ private void HandleAsteriskToken(SyntaxTreeAnalysisContext context, SyntaxToken
...
-            else
-            {
-                SyntaxToken precedingToken = token.GetPreviousToken();
-                precededBySpace = precedingToken.HasTrailingTrivia;
-            }
-
-            bool followedBySpace = token.HasTrailingTrivia;
-            bool lastInLine = followedBySpace && token.TrailingTrivia.Any(SyntaxKind.EndOfLineTrivia);
+            bool firstInLine = token.IsFirstInLine();
+            bool precededBySpace = firstInLine ? true : token.IsPrecededBySpace();
```
Comment (vweijsters) [inline StyleCop.Analyzers/StyleCop.Analyzers/SpacingRules/SA1023DereferenceAndAccessOfSymbolsMustBeSpacedCorrectly.cs]: You could also write this like `precededBySpace = firstInLine || token.IsPrecededBySpace();`

**Comment by vweijsters** [comment]
:+1: Good effort. I made a few minor remarks.

**Comment by Maxwe11** [comment]
updated

File: StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TokenHelpers.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+
+    /// <summary>
+    /// Provides helper methods to work with token.
+    /// </summary>
+    internal static class TokenHelper
+    {
+        /// <summary>
+        /// Gets a value indicating whether the <paramref name="token"/> first in line.
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/Helpers/TokenHelpers.cs]: Nit: **is** first in line. The *is* is missing in the other summaries too.

**Comment by pdelvo** [comment]
I like the change :+1:

**Pull request merged**